R/utilities-simulation-results.R

```diff
@@ -83,6 +83,75 @@ getOutputValues <- function(simulationResults, quantitiesOrPaths, individualIds
   return(output)
 }
 
+#' Export values that can be plugged directly into the TLF lib (TODO)
+#'
+#' @export
+getOutputValuesTLF <- function(simulationResults, population, quantitiesOrPaths = NULL, individualIds = NULL) {
+  validateIsOfType(simulationResults, SimulationResults)
+  validateIsOfType(population, Population)
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+  validateIsOfType(quantitiesOrPaths, c(Quantity, "character"), nullAllowed = TRUE)
+
+  quantitiesOrPaths <- quantitiesOrPaths %||% simulationResults$allQuantityPaths
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(paths, Quantity)) {
+    paths <- unlist(lapply(paths, function(x) x$consolidatePath))
+  }
+  paths <- unique(paths)
+
+  result <- list()
+
+  if (length(paths) == 0) {
+    return(result)
+  }
+
+  # If no specific individual ids are passed, iterate through all individuals
+  individualIds <- ifNotNull(individualIds, unique(individualIds), simulationResults$allIndividualIds)
+
+  # All time values are equal
+  timeValues <- simulationResults$timeValues
+  valueLength <- length(timeValues)
+  covariateNames <- population$allCovariateNames
+
+  metaData <- list(
+    Time = list(unit = "min", dimension = "Time")
+  )
+
+  individualPropertiesCache <- Cache$new()
+  # create a cache of all indivdual values that are constant independent from the path
+  for (individualId in individualIds) {
+    covariates <- population$covariatesAt(individualId)
+    individualProperties <- list(IndividualId = rep(individualId, valueLength), Time = timeValues)
+
+    for (covariateName in covariateNames) {
+      individualProperties[[covariateName]] <- rep(covariates$valueFor(covariateName), valueLength)
+    }
+
+    individualPropertiesCache$set(toString(individualId), individualProperties)
+  }
+
+  for (path in paths) {
+    quantity <- getQuantity(path, simulationResults$simulation)
+    data <- NULL
+    metaData$Value <- list(unit = quantity$unit, dimension = quantity$dimension)
+    pathColumn <- rep(path, valueLength)
+
+    for (individualId in individualIds) {
+      individualProperties <- individualPropertiesCache$get(toString(individualId))
+      values <- simulationResults$getValuesForIndividual(path, individualId)
+      individualData <- data.frame(individualProperties, Path = pathColumn, Value = values)
+      data <- rbind(data, individualData)

```

Comment:
@PavelBal @pchelle @abdullahhamadeh Ok this is the line that brings the performance down. How to rewrite?

Reply:
ok clearly rbind in for loop is not good. Easy to rewrite though: )

Reply:
https://stackoverflow.com/questions/26843861/replace-rbind-in-for-loop-with-lapply-2nd-circle-of-hell